### PR TITLE
Fixed #33107 -- Fixed import_string() crash on not fully initialized modules.

### PR DESCRIPTION
**Pull request by felixxm**
Regression in ecf87ad513fd8af6e4a6093ed918723a7d88d5ca.

Thanks Collin Anderson for the report.

ticket-33107

### CODE REVIEW

**Comment by felixxm** [comment]
@collinanderson Can confirm that it works for you?

**Comment by collinanderson** [comment]
@felixxm it's hard to reproduce, but I've applied the patch to my dev copy and haven't seen any issues. Thanks!

**Comment by collinanderson** [comment]
Actually I figured out I can just put `time.sleep(5)` at the top-level of my custom auth backend to reproduce the issue, and this this patch does fix it. Thanks!

File: django/utils/module_loading.py [hunk 1]
```diff
@@ -7,7 +7,11 @@
...
 def cached_import(module_path, class_name):
     modules = sys.modules
-    if module_path not in modules:
+    if module_path not in modules or (
+        # Module is not fully initialize.
+        getattr(modules[module_path], '__spec__', None) is not None and
+        getattr(modules[module_path].__spec__, '_initializing', False) is True
+    ):
         import_module(module_path)
     return getattr(modules[module_path], class_name)
```
Comment (ngnpope) [inline django/utils/module_loading.py:16]: I'm not sure if we can decruft this a bit?
```diff
def cached_import(module_path, class_name):
    # Check whether module is loaded and fully initialized.
    if not (
        module := sys.modules.get(module_path) and
        spec := getattr(module, '__spec__', None) and
        getattr(spec, '_initializing', False) is False
    ):
        module = import_module(module_path)
    return getattr(module, class_name)
```

You'll have to double check this - just knocked it up quickly.

Comment (felixxm) [reply django/utils/module_loading.py:16]: > I'm not sure if we can decruft this a bit?

It raises a `SyntaxError`.

File: django/utils/module_loading.py [hunk 2]
```diff
@@ -7,7 +7,11 @@
 
 def cached_import(module_path, class_name):
     modules = sys.modules
-    if module_path not in modules:
+    if module_path not in modules or (
+        # Module is not fully initialize.
```
[the remark this answers is not in the document]
Comment (felixxm) [reply django/utils/module_loading.py:11]: Typo, fixed.

**Review by smithdc1 — APPROVED** [review]

**Comment by felixxm** [comment]
@smithdc1 Thanks :+1:

**Pull request merged**